libslide/src/math/fraction.rs

```diff
@@ -0,0 +1,133 @@
+#![allow(unused)]
+
+/// Represents a fraction, consisting of a numerator and denominator.
+type Fraction = (/* num */ i64, /* den */ u64);
+
+/// Converts a decimal number to its irreducible fractional representation by walking the
+/// [Stern-Brocot tree](https://en.wikipedia.org/wiki/Stern%E2%80%93Brocot_tree). This is
+/// equivalent to a binary search of the [Farey sequence](https://en.wikipedia.org/wiki/Farey_sequence).
+///
+/// ## Algorithm
+///
+/// Say we are looking for a number A, where A = Whole + Decimal, Whole ∈ W, and Decimal ∈ [0, 1).
+/// Then if we can find T and B such that Decimal = T / B,
+///      A = Whole + Decimal = (Whole * B) / B + (T / B)
+///        = (Whole * B + T) / B
+///
+/// To find T and B we perform a binary search of irreducible fractions starting with the range [0, 1].
+/// On each iteration of the search, we compute the mediant of the range.
+///
+/// > _mediant_(A / B, C / D) = (A + C) / (B + D)
+/// >
+/// > The _mediant_ of two fractions is guaranteed to be between those two fractions, though is
+/// > **not** always the _median_ of those two fractions.
+/// >
+/// > In the Stern-Brocot tree, the _mediant_ of two sibling tree nodes is an irreducible fraction.
+///      
+/// ```text
+///                       * ~ T / B
+/// 0 / 1 ----------------------|---------------------- 1 / 1
+///                           1 / 2 ~ mediant
+/// ```
+///
+/// We then bisect the search to the medaint-bound range the decimal we are interested in.
+///
+/// ```text
+///                                    * ~ T / B
+/// 0 / 1 -----------------------------|--------------- 1 / 2
+///                                  1 / 3 ~ mediant
+/// ```
+///
+/// We stop when the mediant is equivalent to the decimal searched for.
+///
+/// ## Failure
+///
+/// This algorithm is iterative and convergent, but terminates after `max_iter` iterations. If a
+/// fractional representation equivalent to the decimal at floating point precision cannot be found
+/// after `max_iter`, nothing is returned.
+///
+/// ## Examples
+///
+/// ```ignore
+/// assert_eq!(dec2frac(0.5, 10), Some((1, 2)))
+/// assert_eq!(dec2frac(3.14159265358979323, 10), None)
+/// assert_eq!(dec2frac(-3.142857142857142857, 100_000), Some((-22, 7)))
+/// ```
+pub fn dec2frac(mut num: f64, max_iter: u64) -> Option<Fraction> {
+    let coeff = if num < 0. { -1 } else { 1 };
```

Comment:
use result instead of option.

Reply:
what would be the error type of the result?

Reply:
it should explain to the user why the program failed. Returning none is confusing and I don't know where my expression failed

Reply:
Err("Value {x} could not be converted to decmial")

Reply:
Does this look good?